### ISSUE TITLE
manual_flatten interaction with nested matches could better

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#6776]
I tried this code: manual_flatten

```rust
fn main() {
    let x: Vec<Option<(i32, Option<i32>)>> = vec![Some((4, Some(0)))];
    for n in x.iter() {
        if let Some((_, Some(n))) = n {
            println!("{}", n);
        }
    }
}

```
Clippy warns here:
````
warning: unnecessary `if let` since only the `Some` variant of the iterator element is used
 --> bad.rs:3:5
  |
3 |       for n in x.iter() {
  |       ^        -------- help: try: `x.iter().flatten()`
  |  _____|
  | |
4 | |         if let Some((_, Some(n))) = n {
5 | |             println!("{}", n);
6 | |         }
7 | |     }
  | |_____^
  |
  = note: `#[warn(clippy::manual_flatten)]` on by default
help: ...and remove the `if let` statement in the for loop
 --> bad.rs:4:9
  |
4 | /         if let Some((_, Some(n))) = n {
5 | |             println!("{}", n);
6 | |         }
  | |_________^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_flatten
````

but we cannot remove the entire `if let` because `.flatten()` only peels off the outer `Some()`.

The "fixed" code would be
````rust
fn main() {
    let x: Vec<Option<(i32, Option<i32>)>> = vec![Some((4, Some(0)))];
    for n in x.iter().flatten() {
        if let (_, Some(n)) = n {
            println!("{}", n);
        }
    }
}
````
but despite what the lint suggestion said, we still need the `if let`!

**Comment by Swatinem** [comment]
After hitting this with the recent update, I have a suggestion to improve this:

For *irrefutable* patterns, suggest moving the pattern into the for header like so:

```
// before:
for n in &x {
  if let Ok((a, b, _)) = n => {}
}
// after:
for (a, b, _) = x.iter().flatten() {
}
```

However for refutable matches, especially if matching on enum variants, I’m not sure this rule actually improves the code, and would rather like to disable it.
Especially since it makes the for header harder to read. Instead of `for n in x`, you suddenly have `for n in x.into_iter().flatten()` which is a lot more to type, and definitely longer than the `Ok()` in the unavoidable `if let` block.
